jaeger-core/src/main/java/com/uber/jaeger/Tracer.java

```diff
@@ -276,25 +311,56 @@ boolean isRpcServer() {
       return Tags.SPAN_KIND_SERVER.equals(tags.get(Tags.SPAN_KIND.getKey()));
     }
 
-    @Override
-    public io.opentracing.Span start() {
-      // find preferredParent, it is first childOf
-      Reference preferredParent = references.isEmpty() ? null : references.get(0);
+    private SpanContext preferredParent() {
+      if (parentContext != null) {
+        return parentContext;
+      }
+
+      Reference preferredReference = references.get(0);
       for (Reference reference: references) {
         if (References.CHILD_OF.equals(reference.getType())
-            && !References.CHILD_OF.equals(preferredParent.getType())) {
-          preferredParent = reference;
+            && !References.CHILD_OF.equals(preferredReference.getType())) {
+          preferredReference = reference;
           break;
         }
       }
+      return preferredReference.getSpanContext();
+    }
+
+    private boolean isSampled() {
```

Comment:
again, all this complexity seems to be coming from unnecessary dichotomy between `parentContext` and `references`. If we define `parentContext` to be
  * the first CHILD_OF reference, or
  * the first reference of any type, or
  * null
then we never need to look at `references` for all other values like isSampled, getDebugId, etc.

The ONLY change we would need is
```
   if parentContext != null {
      excludePrimaryChildOf = references.filter(r => r.refContext() != parentContext || r.refType != CHILD_OF)
      if !excludePrimaryChildOf.isEmpty {
          span.references = excludePrimaryChildOf
      }
   } else {
      span.references = references
   }
```